scripts/Plotter.py

```diff
@@ -0,0 +1,241 @@
+
+        plt.xticks(d_list)
+
+        plt.xlabel('Devices')
+        plt.ylabel('SHP Speedup')
+        plt.title(f"Speedup for vec {vec_size} and {reps} reps")
+        plt.legend()
+
+        plt.savefig(f'{self.plots_path_shp}/shp_speedup.png', format='png')
+
+    def _get_bench_vals(self, f):
+        data = json.load(f)
+        benchmarks = data["benchmarks"]
+        copy = round(benchmarks[0]["bytes_per_second"] / 1e9, 2)
+        scale = round(benchmarks[1]["bytes_per_second"] / 1e9, 2)
+        add = round(benchmarks[2]["bytes_per_second"] / 1e9, 2)
+        triad = round(benchmarks[3]["bytes_per_second"] / 1e9, 2)
+
+        self.vec_size = int(data["context"]["default_vector_size"])
+        self.reps = int(data["context"]["default_repetitions"])
+        return BenchResults(copy, scale, add, triad)
+
+    def load_bandwidth(self, mhp_shp: str = "shp"):
+        self.shp_bandwidth_results = dict()
+        self.mhp_bandwidth_results = dict()
+        d_added = set()
+        n_added = set()
+
+        # load shp
+        for filename in os.listdir(self.bench_path_shp):
+            d = Plotter._get_dev_num(filename)
+            if d in d_added or d is None or 'json' not in filename:
+                continue
+            d_added.add(d)
+            with open(os.path.join(self.bench_path_shp, filename), 'r') as f:
+                bench_result = self._get_bench_vals(f)
+                self.shp_bandwidth_results[d] = bench_result
+
+        self.shp_bandwidth_results = dict(
+            sorted(self.shp_bandwidth_results.items())
+        )
+        self.d_list = list(d_added)
+
+        # load mhp
+        for filename in os.listdir(self.bench_path_mhp):
+            n = Plotter._get_mpi_proc_num(filename)
+            if n in n_added or n is None or 'json' not in filename:
+                continue
+            n_added.add(n)
+            with open(os.path.join(self.bench_path_mhp, filename), 'r') as f:
+                bench_result = self._get_bench_vals(f)
+                self.mhp_bandwidth_results[n] = bench_result
+
+        self.mhp_bandwidth_results = dict(
+            sorted(self.mhp_bandwidth_results.items())
+        )
+        self.n_list = list(n_added)
+
+    def _extract_bench_values(self, shp_mhp: str = "shp"):
+        if shp_mhp == "shp":
+            results = self.shp_bandwidth_results
+        else:
+            results = self.mhp_bandwidth_results
+
+        copy, scale, add, triad = [], [], [], []
+        for result in results.values():
+            copy.append(result.copy)
+            scale.append(result.scale)
+            add.append(result.add)
+            triad.append(result.triad)
+
+        return copy, scale, add, triad
+
+    def plot_bandwidth_shp(self):
+        """
+        Plots bandwidths for every number of devices from 1 to n (shp)
+        """
+        copy, scale, add, triad = self._extract_bench_values("shp")
+
+        plt.plot(self.d_list, copy, label='Stream_Copy/bandwidth')
```

Comment:
It's usually a good idea to create figures and axes as needed. So something like
```python
fig = plt.figure(figsize=(12, 6))
ax = fig.add_subplot()
# fig, ax = plt.subplots(figsize=(12, 6))  # shorter way
ax.plot(...)
ax.set_xlabel(...)
plt.savefig(...)
plt.close(fig)  # too many open figures can cause issues
```

Reply:
Done